### ISSUE TITLE
Deadlock involving model.fmut and folder.watchMut

### ISSUE THREAD

**Issue by AudriusButkevicius** [syncthing/syncthing#5929]
Still a thing:
https://gist.github.com/AudriusButkevicius/2bb933a56882ec00706127c7daa0fd7f

**Comment by imsodin** [comment]
```
goroutine 134160 [chan receive, 87 minutes]:
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).internalClose.func1()
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:907 +0x1de
sync.(*Once).Do(0xc01df00588, 0xc000051ef0)
	/usr/local/go/src/sync/once.go:44 +0xb3
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).internalClose(0xc01df004e0, 0x1026100, 0xc00003f060)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:894 +0x6f
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).pingReceiver(0xc01df004e0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:953 +0x3bb
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:260 +0xcb
```

is blocking on

```
		<-c.dispatcherLoopStopped
```

due to `dispatcherLoop` blocking at

```
goroutine 134157 [semacquire, 95 minutes]:
sync.runtime_SemacquireMutex(0xc000bb0384, 0xc03d629200)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000bb0380)
	/usr/local/go/src/sync/mutex.go:134 +0x109
sync.(*RWMutex).Lock(0xc000bb0380)
	/usr/local/go/src/sync/rwmutex.go:93 +0x2d
github.com/syncthing/syncthing/lib/model.(*model).ClusterConfig(0xc0000cadc0, 0xe593b1ce2fe0fbc0, 0x854a417d6b2e05d8, 0xd5b758df0455eb37, 0x9e8a0a4ddaaf99a4, 0xc003b42c00, 0x2, 0x2)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/model.go:1057 +0x45b
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).dispatcherLoop(0xc01df004e0, 0x0, 0x0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:407 +0x116e
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start.func1(0xc01df004e0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:255 +0x2b
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/protocol/protocol.go:254 +0x65
```

which is trying to acquire `fmut` since like forever.

The problem is not a protocol connection deadlock, it's lock contention in the model. There's tons of routines waiting for that lock for over an hour. I was unable to determine which goroutine in the trace is holding `fmut`. In general it would probably be good to have either some kind of priorisation or time-based ordering on that lock (such that requests can't block more important stuff like cluster configs or index updates).

**Comment by calmh** [comment]
We shouldn’t be holding locks over operations that can block, ideally. Maybe some of this can be improved by moving stuff out of model into the folder runners or similar, where things live in their own instance and can run concurrently.

**Comment by imsodin** [comment]
Should we close (and tag as skip-changelog) as it isn't a conn deadlock or repurpose as a general "fight the model lock contention"-issue (where #5943 might have already done some good, I hope Moisie on the forum can give feedback on that)?

**Comment by AudriusButkevicius** [comment]
I don't think 87 minutes of model lock contention is possible. Go mutexes are fair, so the problem is somewhere else

**Comment by AudriusButkevicius** [comment]
fmut is held by:
```
goroutine 114 [semacquire, 100 minutes]:
sync.runtime_SemacquireMutex(0xc0036a54b4, 0x9f3e8efe6aab2200)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0036a54b0)
	/usr/local/go/src/sync/mutex.go:134 +0x109
github.com/syncthing/syncthing/lib/model.(*folder).WatchError(0xc0000cfb80, 0x0, 0x0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:558 +0x46
github.com/syncthing/syncthing/lib/model.(*model).WatchError(0xc0000cadc0, 0xc0004039c0, 0xb, 0x0, 0x0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/model.go:2225 +0x116
github.com/syncthing/syncthing/lib/model.(*folderSummaryService).Summary(0xc0000abcc0, 0xc0004039c0, 0xb, 0x50, 0x458ca0, 0x4bde9e)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder_summary.go:135 +0x12d2
github.com/syncthing/syncthing/lib/model.(*folderSummaryService).sendSummary(0xc0000abcc0, 0xc0004039c0, 0xb)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder_summary.go:279 +0x5d
github.com/syncthing/syncthing/lib/model.(*folderSummaryService).calculateSummaries(0xc0000abcc0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder_summary.go:243 +0x141
github.com/syncthing/syncthing/lib/model.serviceFunc.Serve(0xc0000f8db0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder_summary.go:311 +0x25
github.com/thejerf/suture.(*Supervisor).runService.func1(0xc0007c04b0, 0xc000000001, 0x1030ca0, 0xc0000f8db0)
	/opt/tcagent/syncthing-3-work/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:600 +0x47
created by github.com/thejerf/suture.(*Supervisor).runService
	/opt/tcagent/syncthing-3-work/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:588 +0x5b
```

which tries to acquire watchMut, which is held by:

```
goroutine 168 [chan send, 123 minutes]:
github.com/syncthing/syncthing/lib/model.(*folder).Delay(...)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:254
github.com/syncthing/syncthing/lib/model.(*folder).scanOnWatchErr(0xc0000cfb80)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:690 +0xb2
github.com/syncthing/syncthing/lib/model.(*folder).monitorWatch(0xc0000cfb80, 0x1038ee0, 0xc003561140)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:628 +0x2e5
created by github.com/syncthing/syncthing/lib/model.(*folder).startWatch
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:610 +0x10a
```

which is not reading off the channel because it's trying to kick off a scan:

```
goroutine 26 [semacquire, 63 minutes]:
sync.runtime_SemacquireMutex(0xc000bb0384, 0x0)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000bb0380)
	/usr/local/go/src/sync/mutex.go:134 +0x109
sync.(*RWMutex).Lock(0xc000bb0380)
	/usr/local/go/src/sync/rwmutex.go:93 +0x2d
github.com/syncthing/syncthing/lib/model.(*model).numHashers(0xc0000cadc0, 0xc0001157d0, 0xb, 0x1)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/model.go:2111 +0x52
github.com/syncthing/syncthing/lib/model.(*folder).scanSubdirs(0xc0000cf600, 0x0, 0x0, 0x0, 0x0, 0x0)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:349 +0x302
github.com/syncthing/syncthing/lib/model.(*folder).scanTimerFired(0xc0000cf600)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:541 +0x54
github.com/syncthing/syncthing/lib/model.(*folder).Serve(0xc0000cf600)
	/opt/tcagent/syncthing-3-work/174e136266f8a219/lib/model/folder.go:178 +0x7c9
github.com/thejerf/suture.(*Supervisor).runService.func1(0xc0001984b0, 0xc000000002, 0x7f1b00694e88, 0xc0000cf600)
	/opt/tcagent/syncthing-3-work/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:600 +0x47
created by github.com/thejerf/suture.(*Supervisor).runService
	/opt/tcagent/syncthing-3-work/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:588 +0x5b
```

which requires acuiring fmut.

**Comment by AudriusButkevicius** [comment]
I think for single value things, like watch error, we should start using atomic.Value.

I am not sure if go sucks with it's "concurrency is awesome", or is our code so shit that we constantly have these problems

**Comment by imsodin** [comment]
> I am not sure if go sucks with it's "concurrency is awesome", or is our code so shit that we constantly have these problems

Out code is probably a tiny (or not so tiny) bit complex/overly nested, but in general I think the main problem is me learning by making mistakes - if you look at the blame of these code parts the verdict is pretty clear. In hindsight this looks so very blatantly needlessly dangerous: In two places involved in the deadlock you detected I am holding locks while calling complex methods on the protected objects, instead of releasing the lock after getting the object and then calling the methods. If I hadn't done that mistake in just one place, there wouldn't be a deadlock :P

**Comment by AudriusButkevicius** [comment]
I am somehow not sure that this is the only skeleton in the closet in the recent release, as watch errors are not that common, and we've had atleast a few folks not happy with this.

I think our "fmut held for too long" panic is reporting these to sentry, yet all the different flavours of this get folded under a single message.

**Comment by imsodin** [comment]
You might very well be right again, there are quite a few instances with the same problematic pattern: https://github.com/syncthing/syncthing/pull/5944

**Comment by imsodin** [comment]
The deadlock described in https://github.com/syncthing/syncthing/issues/5929#issuecomment-520578153 is closed in #5944 and I don't think there's anything else that's actionable here.
